File: app/src/main/java/cz/civilizacehra/cipherbreaker/NumberAnalysis.kt
package cz.civilizacehra.cipherbreaker

import kotlin.math.sqrt

internal object NumberAnalysis {
    const val ROMAN_NUMERALS = "roman numerals"
    const val PRIME_FACTORS = "prime factors"

    // Above this the numeral is a wall of M characters, so it is not worth showing
    private val MAX_ROMAN_NUMERAL = 100000.toULong()

    private val ROMAN_SYMBOLS = listOf(
            1000.toULong() to "M", 900.toULong() to "CM", 500.toULong() to "D",
            400.toULong() to "CD", 100.toULong() to "C", 90.toULong() to "XC",
            50.toULong() to "L", 40.toULong() to "XL", 10.toULong() to "X",
            9.toULong() to "IX", 5.toULong() to "V", 4.toULong() to "IV",
            1.toULong() to "I")
    // Only canonical numerals: IV/IX/XL/XC/CD/CM are the only subtractive pairs,
    // I/X/C repeat at most three times and V/L/D at most once. Thousands are left
    // unbounded, because values above MMM have no other plain text notation.
    private val ROMAN_PATTERN =
            Regex("M*(CM|CD|D?C{0,3})(XC|XL|L?X{0,3})(IX|IV|V?I{0,3})")

    private val ROMAN_VALUES = mapOf(
            'I' to 1.toULong(), 'V' to 5.toULong(), 'X' to 10.toULong(), 'L' to 50.toULong(),
            'C' to 100.toULong(), 'D' to 500.toULong(), 'M' to 1000.toULong())

    // Input type is either "roman numerals" or "base-N".
    fun parseNumber(input: String, inputType: String): ULong? {
        if (inputType == ROMAN_NUMERALS) {
            return parseRomanNumeral(input)
        }
        val radix = inputType.removePrefix("base-").toIntOrNull() ?: return null
        return input.toULongOrNull(radix)
    }

    fun formatInBase(number: ULong, radix: Int): String {
        val digits = number.toString(radix)
        // Hexadecimal reads better in capitals, and matches what the hex keyboard types
        return if (radix > 10) digits.uppercase() else digits
    }

    fun formatRomanNumeral(number: ULong): String? {
        if (number < 1.toULong() || number > MAX_ROMAN_NUMERAL) {
            return null
        }

        var remainder = number
        val numeral = StringBuilder()
        // Greedily take the largest symbol that still fits, subtractive pairs included
        for ((value, symbol) in ROMAN_SYMBOLS) {
            while (remainder >= value) {
                numeral.append(symbol)
                remainder -= value
            }
        }
        return numeral.toString()
    }

    fun parseRomanNumeral(input: String): ULong? {
        val numeral = input.uppercase()
        if (numeral.isEmpty() || !ROMAN_PATTERN.matches(numeral)) {
            return null
        }

        var total = 0.toULong()
        var previous = 0.toULong()
        // Walk right to left: a numeral smaller than the one to its right is subtracted
        for (c in numeral.reversed()) {
            val value = ROMAN_VALUES.getValue(c)
            if (value < previous) {
                total -= value
            } else {
                total += value
                previous = value
            }
        }
        return total
    }

    fun factorNumber(number: ULong): ArrayList<ULong> {
        val factors: ArrayList<ULong> = arrayListOf()
        if (number < 2.toULong()) {
            return factors
        }
        var n = number
        val squareRoot = sqrt(number.toDouble()).toULong()

        // At first check for divisibility by 2. add it in arr till it is divisible
        while (n % 2u == 0.toULong()) {
            factors.add(2u)
            n /= 2u
        }

        // Run loop from 3 to square root of n. Check for divisibility by i.
        // Add i in arr till it is divisible by i.
        for (i in 3.toULong()..squareRoot step 2) {
            while (n % i == 0.toULong()) {
                factors.add(i)
                n /= i
            }
            if (n < i) {
                break
            }
        }

        // If n is a prime number greater than 2.
        if (n > 2u) {
            factors.add(n)
        }
        return factors
    }
}

File: pwa/src/logic/number-analysis.ts
// Parsing, formatting and factoring of unsigned 64-bit numbers, which are bigint here as they do
// not fit into a JS number. Port of NumberAnalysis.kt

export const ROMAN_NUMERALS = 'roman numerals';
export const PRIME_FACTORS = 'prime factors';

export const ULONG_MAX = (1n << 64n) - 1n;

// Above this the numeral is a wall of M characters, so it is not worth showing
const MAX_ROMAN_NUMERAL = 100000n;

const ROMAN_SYMBOLS: [bigint, string][] = [
  [1000n, 'M'], [900n, 'CM'], [500n, 'D'], [400n, 'CD'], [100n, 'C'], [90n, 'XC'],
  [50n, 'L'], [40n, 'XL'], [10n, 'X'], [9n, 'IX'], [5n, 'V'], [4n, 'IV'], [1n, 'I'],
];
// Only canonical numerals: IV/IX/XL/XC/CD/CM are the only subtractive pairs,
// I/X/C repeat at most three times and V/L/D at most once. Thousands are left
// unbounded, because values above MMM have no other plain text notation.
const ROMAN_PATTERN = /^M*(CM|CD|D?C{0,3})(XC|XL|L?X{0,3})(IX|IV|V?I{0,3})$/;

const ROMAN_VALUES: Record<string, bigint> = {
  I: 1n, V: 5n, X: 10n, L: 50n, C: 100n, D: 500n, M: 1000n,
};

const DIGITS = '0123456789abcdefghijklmnopqrstuvwxyz';

/** Parses digits of given radix as an unsigned 64-bit number, null when they are not one. */
export function parseULong(input: string, radix: number): bigint | null {
  if (input === '' || radix < 2 || radix > 36) return null;
  let value = 0n;
  const big = BigInt(radix);
  for (const c of input.toLowerCase()) {
    const digit = DIGITS.indexOf(c);
    if (digit < 0 || digit >= radix) return null;
    value = value * big + BigInt(digit);
    if (value > ULONG_MAX) return null;
  }
  return value;
}

// Input type is either "roman numerals" or "base-N".
export function parseNumber(input: string, inputType: string): bigint | null {
  if (inputType === ROMAN_NUMERALS) return parseRomanNumeral(input);
  if (!inputType.startsWith('base-')) return null;
  const radix = parseInt(inputType.slice('base-'.length), 10);
  if (!Number.isFinite(radix)) return null;
  return parseULong(input, radix);
}

export function formatInBase(number: bigint, radix: number): string {
  const digits = number.toString(radix);
  // Hexadecimal reads better in capitals, and matches what the hex keyboard types
  return radix > 10 ? digits.toUpperCase() : digits;
}

export function formatRomanNumeral(number: bigint): string | null {
  if (number < 1n || number > MAX_ROMAN_NUMERAL) return null;
  let remainder = number;
  let numeral = '';
  // Greedily take the largest symbol that still fits, subtractive pairs included
  for (const [value, symbol] of ROMAN_SYMBOLS) {
    while (remainder >= value) {
      numeral += symbol;
      remainder -= value;
    }
  }
  return numeral;
}

export function parseRomanNumeral(input: string): bigint | null {
  const numeral = input.toUpperCase();
  if (numeral === '' || !ROMAN_PATTERN.test(numeral)) return null;
  let total = 0n;
  let previous = 0n;
  // Walk right to left: a numeral smaller than the one to its right is subtracted
  for (let i = numeral.length - 1; i >= 0; i--) {
    const value = ROMAN_VALUES[numeral[i]];
    if (value < previous) {
      total -= value;
    } else {
      total += value;
      previous = value;
    }
  }
  return total;
}

/** Prime factors in ascending order, with repetitions. Numbers below 2 have none. */
export function factorNumber(number: bigint): bigint[] {
  const factors: bigint[] = [];
  if (number < 2n) return factors;
  let n = number;
  while (n % 2n === 0n) {
    factors.push(2n);
    n /= 2n;
  }
  // Odd divisors up to the square root of what is left
  for (let i = 3n; i * i <= n; i += 2n) {
    while (n % i === 0n) {
      factors.push(i);
      n /= i;
    }
  }
  // What is left is a prime greater than 2
  if (n > 2n) factors.push(n);
  return factors;
}
